coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+	// Format frequency label
+	var frequencyLabel string
+	if frequencyDays == 7 {
+		frequencyLabel = "week"
+	} else {
+		var plural string
+		if frequencyDays > 1 {
+			plural = "s"
+		}
+		frequencyLabel = fmt.Sprintf("%d day%s", frequencyDays, plural)
+	}
+
+	// Sorting order: template_version_name ASC, workspace build number DESC
+	sort.Slice(failedBuilds, func(i, j int) bool {
+		if failedBuilds[i].TemplateVersionName != failedBuilds[j].TemplateVersionName {
+			return failedBuilds[i].TemplateVersionName < failedBuilds[j].TemplateVersionName
+		}
+		return failedBuilds[i].WorkspaceBuildNumber > failedBuilds[j].WorkspaceBuildNumber
+	})
+
+	// Build notification model for template versions and failed workspace builds
+	templateVersions := []map[string]any{}
+	for _, failedBuild := range failedBuilds {
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
+			templateVersions = append(templateVersions, map[string]any{
+				"template_version_name": failedBuild.TemplateVersionName,
+				"failed_count":          1,
+				"failed_builds": []map[string]any{
+					{
+						"workspace_owner_username": failedBuild.WorkspaceOwnerUsername,
+						"workspace_name":           failedBuild.WorkspaceName,
+						"build_number":             failedBuild.WorkspaceBuildNumber,
+					},
+				},
+			})
+			continue
+		}
+
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
+		builds := templateVersions[c-1]["failed_builds"].([]map[string]any)
+		builds = append(builds, map[string]any{
+			"workspace_owner_username": failedBuild.WorkspaceOwnerUsername,
+			"workspace_name":           failedBuild.WorkspaceName,
+			"build_number":             failedBuild.WorkspaceBuildNumber,
+		})
+		templateVersions[c-1]["failed_builds"] = builds
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
+		templateVersions[c-1]["failed_count"] = templateVersions[c-1]["failed_count"].(int) + 1
+	}
+
+	return map[string]any{
+		"failed_builds":     stats.FailedBuilds,
+		"total_builds":      stats.TotalBuilds,
+		"report_frequency":  frequencyLabel,
+		"template_versions": templateVersions,
+	}
+}
+
+func findTemplateAdmins(ctx context.Context, db database.Store, stats database.GetWorkspaceBuildStatsByTemplatesRow) ([]database.GetUsersRow, error) {
+	users, err := db.GetUsers(ctx, database.GetUsersParams{
+		RbacRole: []string{codersdk.RoleTemplateAdmin},
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("unable to fetch template admins: %w", err)
+	}
+
+	usersByIDs := map[uuid.UUID]database.GetUsersRow{}
+	var userIDs []uuid.UUID
+	for _, user := range users {
+		usersByIDs[user.ID] = user
+		userIDs = append(userIDs, user.ID)
+	}
+
+	var templateAdmins []database.GetUsersRow
+	if len(userIDs) > 0 {
```

Comment:
Nit: Invert and return early
The go idiom is that the most unintended code should be the happy path.